assets/js/utils/retry.ts

```diff
@@ -0,0 +1,125 @@
+export interface RetryParams {
+  /**
+   * Maximum number of attempts to retry the operation. The first attempt counts
+   * too, so setting this to 1 is equivalent to no retries.
+   */
+  maxAttempts?: number;
+
+  /**
+   * Initial delay for the first retry. Subsequent retries will be exponentially
+   * delayed up to `maxDelayMs`.
+   */
+  minDelayMs?: number;
+
+  /**
+   * Max value a delay can reach. This is useful to avoid unreasonably long
+   * delays that can be reached at a larger number of retries where the delay
+   * grows exponentially very fast.
+   */
+  maxDelayMs?: number;
+
+  /**
+   * If present determines if the error should be retried or immediately re-thrown.
+   * All errors that aren't instances of `Error` are considered non-retryable.
+   */
+  isRetryable?(error: Error): boolean;
+
+  /**
+   * Human-readable message to identify the operation being retried. By default
+   * the function name is used.
+   */
+  label?: string;
+}
+
+export type RetryFunc<R = void> = (attempt: number, nextDelayMs?: number) => Promise<R>;
+
+/**
+ * Retry an async operation with exponential backoff and jitter.
+ *
+ * The callback receives the current attempt number and the delay before the
+ * next attempt in case the current attempt fails. The next delay may be
+ * `undefined` if this is the last attempt and no further retries will be scheduled.
+ *
+ * This is based on the following AWS paper:
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+ */
+export async function retry<R>(func: RetryFunc<R>, params?: RetryParams): Promise<R> {
+  const maxAttempts = params?.maxAttempts ?? 3;
+
+  if (maxAttempts < 1) {
+    throw new Error(`Invalid 'maxAttempts' for retry: ${maxAttempts}`);
+  }
+
+  const minDelayMs = params?.minDelayMs ?? 200;
+
+  if (minDelayMs < 0) {
+    throw new Error(`Invalid 'minDelayMs' for retry: ${minDelayMs}`);
+  }
+
+  const maxDelayMs = params?.maxDelayMs ?? 1500;
+
+  if (maxDelayMs < minDelayMs) {
+    throw new Error(`Invalid 'maxDelayMs' for retry: ${maxDelayMs}, 'minDelayMs' is ${minDelayMs}`);
+  }
+
+  const label = params?.label || func.name || '{unnamed routine}';
+
+  const backoffExponent = 2;
+
+  let attempt = 1;
+  let nextDelayMs = minDelayMs;
+
+  while (true) {
+    const hasNextAttempts = attempt < maxAttempts;
+
+    try {
+      // XXX: an `await` is important in this block to make sure the exception is caught
+      // in this scope. Doing a `return func()` would be a big mistake, so don't try
+      // to "refactor" that!
+      const result = await func(attempt, hasNextAttempts ? nextDelayMs : undefined);
+      return result;
```

Comment:
Should be equivalent to
```suggestion
      return await func(attempt, hasNextAttempts ? nextDelayMs : undefined);
```
, no?

Reply:
Should be, but check the `XXX` comment on top of it. The extra variable indirection there is intentional to prevent someone from refactoring it from `return await ...` to just `return ...`. Anyway, if you insist, I can accept this suggestion. Also why removing the comment in this suggestion?

Reply:
Edited the suggestion. I would personally consider this more of a nota bene than a XXX

Reply:
Committed